### ISSUE TITLE
Flipping causes inaccurate X-axis coordinates

### ISSUE THREAD

**Issue by Deprecator16** [openseadragon/openseadragon#1695]
Hello,
I came across an issue where flipping the image causes the viewport x-axis coordinates, and inherently the image x coordinates, to be completely inaccurate. When the image is not flipped, the top left of the image is always 0,0. When flipped, I would expect the top left to be either 1,0 since it was flipped. Instead, the viewport x coordinates vary based on where the image is in the viewport and don't seem to be tied directly to the image. 

I think the affects can be seen the best at https://openseadragon.github.io/examples/viewport-coordinates. Flip the image, then move it around and check the viewport and image x coordinates of any point on the image. It will not be accurate.

From what I can tell, the bounds are moving in the opposite direction that the image is moving. For example, using the link above, drag the image all the way to the right. The viewport x coordinate at the left most side will be 0.5, and the right most side will be 1.5, showing that the bounds move in the exact opposite direction.

I'm still looking into the issue, but I think it may be related to the way the controls are inverted when an image is flipped, since it only affects the x-axis. I'm not 100% sure though, I could be completely wrong. I'd be happy to try to fix this if I can be pointed in the right direction.

**Comment by Deprecator16** [comment]
On further investigation, viewport.getCenter() returns the correct value, however viewport.getBoundsNoRotate() does not. I haven't found a solution, but I've tried a couple of things. I replaced this in getBoundsNoRotate():

```````javascript
return new $.Rect(
                center.x - (width / 2.0),
                center.y - (height / 2.0),
                width,
                height
            );
``````````

with:

```````javascript
if (this.flipped) {
            return new $.Rect(
                center.x + (width / 2.0),
                center.y - (height / 2.0),
                width,
                height
            );

} else {
            return new $.Rect(
                center.x - (width / 2.0),
                center.y - (height / 2.0),
                width,
                height
            );
        }
``````````

This provides a rudimentary fix by inverting the x-axis but it still does not fix the issue entirely. So, I made it return a negative width as well. This made the viewport x coordinates consistent, for example, clicking on the same pixel on the image, no matter where it was moved, would still return the same viewport x coordinate. However, this also inverted the tiles, making it harder to assess what was actually happening. I'll continue looking into it. Hopefully this helps pinpoint the issue a little better.

**Comment by iangilman** [comment]
Good point! It looks like while the viewport stores the `flipped` value, it's actually not making any use of it itself, so of course the coordinates aren't right. It looks like the rotation does do the right thing, so we might need to take a cue from how rotation is dealt with in the viewport. Doing this right might require some refactoring to how `flipped` is handled in the tiledImage and drawer.

I imagine the most basic level is getting `pointFromPixel` to do the right thing. Many other things depend on that.

@nelsonn3c wrote the flip feature; any thoughts on this topic?

@Deprecator16 Thank you for looking into this!

**Comment by roof12** [comment]
I tried briefly to fix this, but my attempts had consequences that I didn't understand.

Here is the workaround I use to calculate pixelFromPoint:

```
var p;
if (this._viewer.viewport.getFlip()) {
    p = this._viewer.viewport.pixelFromPoint(new $.Point(1, 0), true);
    p.x = this._viewer.viewport._containerInnerSize.x;
} else {
    p = this._viewer.viewport.pixelFromPoint(new $.Point(0, 0), true);
}
```

**Comment by iangilman** [comment]
@roof12 thank you for looking into it and for sharing your workaround!

If you want to dig into it further, I'd love to hear what the consequences were that you found.

**Comment by nelsonn3c** [comment]
Hello guys,

Sorry for my absence. Some recent changes in my life won't allow me to contribute in a near future.

Thank you for everything, all the support and your precious work/time.

**Comment by iangilman** [comment]
@nelsonn3c Thank you for the update. I hope all is well! Thank you for your contributions :)

**Comment by iangilman** [comment]
Note that @jstaab just made a change in this area with https://github.com/openseadragon/openseadragon/pull/1779. I suspect it doesn't entirely fix this issue, but it may be a step in the right direction (or it might be the whole fix). Anyone who is still interested in this issue, I'd love it if you could try out the latest master to ascertain the status!
